src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -0,0 +1,316 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ODataUtf8JsonWriter.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Text.Encodings.Web;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// Implementation of <see cref="IJsonWriter"/> that is based on
+    /// <see cref="Utf8JsonWriter"/>.
+    /// </summary>
+    internal sealed class ODataUtf8JsonWriter : IJsonWriter, IDisposable
+    {
+        private const int DefaultBufferSize = 16 * 1024;
+        private readonly float bufferFlushThreshold;
+
+        private readonly Stream outputStream;
+        private readonly Stream writeStream;
+        private readonly Utf8JsonWriter writer;
+        private readonly int bufferSize;
+        private readonly bool isIeee754Compatible;
+        private readonly bool leaveStreamOpen;
+        private readonly Encoding outputEncoding;
+        private bool disposed;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUtf8JsonWriter"/>.
+        /// </summary>
+        /// <param name="outputStream">The stream to write to.</param>
+        /// <param name="isIeee754Compatible">When true, longs and decimals will be written out as strings.</param>
+        /// <param name="encoding">The text encoding to write to.</param>
+        /// <param name="bufferSize">The internal buffer size.</param>
+        /// <param name="leaveStreamOpen">Whether to leave the <paramref name="outputStream"/> open when this writer gets disposed.</param>
+        public ODataUtf8JsonWriter(Stream outputStream, bool isIeee754Compatible, Encoding encoding, JavaScriptEncoder encoder = null, int bufferSize = DefaultBufferSize, bool leaveStreamOpen = false)

```

Comment:
Out of curiosity, how did you settle on `false` as the more appropriate default for `leaveStreamOpen`?

Reply:
I found that a number of Stream writer classes in the .NET library that provide a similar argument set it to false by default, e.g. `StreamWriter`, `StreamReader`, `Encoding.CreateTranscodingStream`.

Besides that, my main reason is to require the user to explicitly make the choice to clean up the stream on their own, i.e. "I know what I'm doing, let me dispose this myself". I find that this has the best worst case:
If the user does not want the stream to be auto-disposed, but forgets to set `leaveStreamOpen` to false, then it's likely that the caller will use the stream after the writer has disposed it. In that case they'll likely get an `ObjectDisposedException` at runtime. That's not ideal, but at least it's relatively easy to detect, trace down and fix.

Now on the flip side, if the writer did not dispose the stream by default, and the user does not dispose the stream either (maybe they forgot, or they assume it's done automatically), then they will probably not get any errors. The stream will just not get cleaned up. This might lead to resource leaks or expensive finalizer calls that can go undetected and are harder to trace down compared to a `ObjectDisposedException`. This is the main reason I prefer this as a general go-to pattern for classes that wrap streams.

@corranrogue9 might not agree with this take though 😅